websockets-protocol-binding.md

```diff
@@ -0,0 +1,162 @@
+"SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be
+interpreted as described in [RFC2119][rfc2119].
+
+### 1.2. Relation to WebSockets
+
+This specification does not prescribe rules constraining the use or handling of
+specific [HTTP target resource][rfc7230-section-5-1] to establish the WebSocket
+upgrade.
+
+This specification prescribe rules constraining the [WebSockets
+Subprotocols][rfc6455-section-5-1] in order to reach agreement on the event
+format to use, in order to exchange serialized CloudEvents.
+
+Events are sent as WebSocket messages, serialized using an [event
+format][ce-event-format].
+
+### 1.3. Content Modes
+
+The [CloudEvents specification][ce-message] defines two content modes for
+transferring events: _structured_ and _binary_.
+
+Because of the nature of WebSockets messages, this specification supports only
+_structured_ data mode, hence event metadata attributes and event data are
+placed into the WebSockets messages using an [event format][ce-event-format].
+
+The [event format][ce-event-format] to be used in a full-duplex WebSocket stream
+is agreed during the [handshake](#14-handshake) and cannot change during the
+same stream.
+
+### 1.4. Handshake
+
+The [opening handshake][rfc6455-section-1-3] MUST follow the set of rules
+specified in the [RFC6455][rfc6455-section-4].
+
+In addition, the client MUST include, in the opening handshake, the
+[`Sec-WebSocket-Protocol` header][rfc6455-section-1-9]. The client MUST include
+in this header one or more [CloudEvents subprotocols](#15-cloudevents-subprotocols),
+depending on the subprotocols the client supports.
+
+The server MUST reply with the chosen CloudEvents subprotocol using the
+`Sec-WebSocket-Protocol` header. If the server doesn't support any of the
+subprotocols included in the opening handshake, the server response should not
+contain any `Sec-WebSocket-Protocol` header.
+
+#### 1.4.1 Example
+
+Example client request to begin the opening handshake:
+
+```text
+GET /events HTTP/1.1
+Host: server.example.com
+Upgrade: websocket
+Connection: Upgrade
+Sec-WebSocket-Key: x3JJHMbDL1EzLkh9GBhXDw==
+Sec-WebSocket-Protocol: json.cloudevents, avro.cloudevents
+Sec-WebSocket-Version: 13
+Origin: http://example.com
+```
+
+Example server response:
+
+```text
+HTTP/1.1 101 Switching Protocols
+Upgrade: websocket
+Connection: Upgrade
+Sec-WebSocket-Accept: HSmrc0sMlYUkAGmm5OPpG2HaGWk=
+Sec-WebSocket-Protocol: json.cloudevents
+```
+
+### 1.5. CloudEvents Subprotocols
+
+This specification maps a WebSocket subprotocol to each defined event format in
+the CloudEvents specification, following the guidelines discussed in the
+[RFC6455][rfc6455-section-1-9]:
+
+| Subprotocol        | Event format                     |
+| ------------------ | -------------------------------- |
+| `json.cloudevents` | [JSON event format][json-format] |
+| `avro.cloudevents` | [AVRO event format][avro-format] |
```

Comment:
After looking at the [W3C interface](https://html.spec.whatwg.org/multipage/web-sockets.html#the-websocket-interface) again, it appears that using a subprotocol that includes the event format is indeed the best option we have, even though I still find that iffy. 

[OneM2M seems to have established a precedent](https://www.iana.org/assignments/websocket/websocket.xhtml)  in the IANA registry. I would switch the names to "cloudevents.*". We should reach out to IANA and register at least "cloudevents.json", "cloudevents.avro", "cloudevents.protobuf", and "cloudevents.cbor". 

What's missing for the subprotocol definitions is whether the subprotocol is using binary or text frames. I think we will want to default to use text frames for JSON, but binary frames for Avro and Proto.

Reply:
> What's missing for the subprotocol definitions is whether the subprotocol is using binary or text frames. I think we will want to default to use text frames for JSON, but binary frames for Avro and Proto.

Should we define it in this table?

Reply:
check it out now